Test failure on 32-bit arches

I am attempting to build e-antic 1.0.1 for the Fedora Linux distribution.  The build succeeded on every 64-bit architecture, but tests failed on both 32-bit architectures (i686 and ARM).  In both cases, test-suite.log shows the following:
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
t-cmp is a Catch v2.13.5 host application.
Run with -? for options

-------------------------------------------------------------------------------
Relational Operators with Integers - int
  Implications
-------------------------------------------------------------------------------
renfxx/t-cmp.cpp:48
...............................................................................

renfxx/t-cmp.cpp:76: FAILED:
  REQUIRE( not (a <= b) )
with expansion:
  false

-------------------------------------------------------------------------------
Relational Operators with Integers - long
  Implications
-------------------------------------------------------------------------------
renfxx/t-cmp.cpp:48
...............................................................................

renfxx/t-cmp.cpp:76: FAILED:
  REQUIRE( not (a <= b) )
with expansion:
  false

===============================================================================
test cases:     9 |     7 passed | 2 failed
assertions: 97081 | 97079 passed | 2 failed

FAIL renfxx/t-cmp (exit status: 2)
```

GDB shows that when this test fails, the value of b is -2147483648; i.e., the most negative integer, and the only integer besides zero that is the negation of itself.  Let me know if I can provide any other information.

The dependencies have changed some since July 16, but I still see the issue.  Currently:

- antic 0.2.4
- arb 2.20
- binutils 2.37
- boost 1.76.0
- catch 2.13.5
- flint 2.8.0
- gcc 11.2.1
- glibc 2.34.9000
- mpfr 4.1.0 + post-release patches 1-13
- ntl 11.5.1

The build flags used by Fedora are as follows:
```
-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -mtune=generic -msse2 -mfpmath=sse -mstackrealign -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection
```

Things I have tried, without success:

- Add -fwrapv to the build flags
- Fix this warning:
```
../e-antic/renf_elem_class.hpp:213:48: warning: 'renf_elem_t' violates the C++ One Definition Rule [-Wodr]
  213 |     std::add_lvalue_reference_t<::renf_elem_t> renf_elem_t() const;
      |                                                ^
renf_elem_class.cpp:661:17: note: return value type mismatch
  661 | ::renf_elem_t & renf_elem_class::renf_elem_t() const
      |                 ^
renf_elem_class.cpp:661:17: note: 'renf_elem_t' was previously declared here
```
I changed line 661 of libeantic/srcxx/renf_elem_class.cpp to this, to match the header:
```
std::add_lvalue_reference_t<::renf_elem_t> renf_elem_class::renf_elem_t() const
```
- Disable LTO

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -29,7 +29,8 @@ jobs:
             cc: clang
             cpp: clang-cpp
             cxx: clang++
-            cflags: -Werror -Weverything -Wno-padded -Wno-disabled-macro-expansion -Wno-float-equal -Wno-documentation
+            # We need -Wno-reselved-identifiers since we extend FLINT using their style of having internal functions start with an underscore.
+            cflags: -Werror -Weverything -Wno-padded -Wno-disabled-macro-expansion -Wno-float-equal -Wno-documentation -Wno-reserved-identifier
             cxxflags: -Wno-exit-time-destructors -Wno-undefined-func-template -Wno-global-constructors -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-zero-as-null-pointer-constant -Wno-deprecated-declarations -Wno-shadow-field-in-constructor -Wno-documentation
     steps:
       - uses: actions/checkout@v2

--- a/doc/news/cmpsgn.rst
+++ b/doc/news/cmpsgn.rst
@@ -1,0 +1,12 @@
+**Changed:**
+
+* We do not claim anymore that `_cmp` functions return -1, 0, 1. Instead we
+  only make a statement about the sign of the integer returned. In practice we
+  still return -1, 0, 1 for the time being.
+
+**Fixed:**
+
+* We do not assume anymore that `_cmp` functions return -1, 0, 1. FLINT, GMP,
+  and Arb only guarantee that these functions return a negative, zero, or
+  positive integer. In practice their implementations return -1, 0, 1, however
+  in some obscure case on Fedora/i686, this was not the case.

--- a/libeantic/e-antic/renf_elem.h
+++ b/libeantic/e-antic/renf_elem.h
@@ -144,10 +144,10 @@ LIBEANTIC_API double renf_elem_get_d(renf_elem_t a, renf_t nf, arf_rnd_t rnd);
 
 /// ==* Comparisons *==
 /// All the comparison functions `renf_elem_cmp_…` between two elements `a` and
-/// `b` behave as follows. They return:
-/// * `1` if `a` is greater than `b`
-/// * `0` if `a` and `b` are equal
-/// * `-1` if `a` is smaller than `b`
+/// `b` behave as follows. They return
+/// * a positive integer if `a` is greater than `b`,
+/// * `0` if `a` and `b` are equal, and
+/// * a negative integer if `a` is smaller than `b`.
 ///
 /// If you want to check for equality, use the faster `renf_elem_equal_…` functions.
 LIBEANTIC_API int renf_elem_cmp(renf_elem_t a, renf_elem_t b, renf_t nf);

--- a/libeantic/src/fmpz_poly_extra/bisection_step_arb.c
+++ b/libeantic/src/fmpz_poly_extra/bisection_step_arb.c
@@ -9,7 +9,7 @@
     (at your option) any later version.  See <http://www.gnu.org/licenses/>.
 */
 
-#include "../e-antic/fmpz_poly_extra.h"
+#include "../../e-antic/fmpz_poly_extra.h"
 
 void _fmpz_poly_bisection_step_arf(arf_t l, arf_t r, const fmpz * pol, slong len, int sl, int sr, slong prec)
 {

--- a/libeantic/src/renf_elem/fdiv.c
+++ b/libeantic/src/renf_elem/fdiv.c
@@ -1,5 +1,5 @@
 /*
-    Copyright (C) 2020 Julian Rüth
+    Copyright (C) 2020-2021 Julian Rüth
 
     This file is part of e-antic
 
@@ -12,12 +12,6 @@
 #include <arb.h>
 
 #include "../e-antic/renf_elem.h"
-
-static int sgn(int a) {
-  if (a == 0) return 0;
-  else if (a > 0) return 1;
-  return -1;
-}
 
 void renf_elem_fdiv(fmpz_t a, renf_elem_t b, renf_elem_t c, renf_t nf)
 {
@@ -75,7 +69,11 @@ void renf_elem_fdiv(fmpz_t a, renf_elem_t b, renf_elem_t c, renf_t nf)
             renf_elem_init(ac, nf);
             renf_elem_mul_fmpz(ac, c, a, nf);
 
-            if (sgn(renf_elem_cmp(ac, b, nf)) == renf_elem_sgn(c, nf))
+            int cmp = renf_elem_cmp(ac, b, nf);
+            int csgn = renf_elem_sgn(c, nf);
+
+            // If a > b/c, we need to subtract one from a.
+            if (cmp == csgn)
                 fmpz_add_si(a, a, -1);
 
             renf_elem_clear(ac, nf);

--- a/libeantic/src/renf_elem/fdiv.c
+++ b/libeantic/src/renf_elem/fdiv.c
@@ -13,6 +13,12 @@
 
 #include "../e-antic/renf_elem.h"
 
+static int sgn(int a) {
+  if (a == 0) return 0;
+  else if (a > 0) return 1;
+  return -1;
+}
+
 void renf_elem_fdiv(fmpz_t a, renf_elem_t b, renf_elem_t c, renf_t nf)
 {
     if (!renf_elem_is_rational(c, nf))
@@ -69,7 +75,7 @@ void renf_elem_fdiv(fmpz_t a, renf_elem_t b, renf_elem_t c, renf_t nf)
             renf_elem_init(ac, nf);
             renf_elem_mul_fmpz(ac, c, a, nf);
 
-            if (renf_elem_cmp(ac, b, nf) == renf_elem_sgn(c, nf))
+            if (sgn(renf_elem_cmp(ac, b, nf)) == renf_elem_sgn(c, nf))
                 fmpz_add_si(a, a, -1);
 
             renf_elem_clear(ac, nf);

--- a/libeantic/src/renf_elem/get_d.c
+++ b/libeantic/src/renf_elem/get_d.c
@@ -30,7 +30,7 @@ static int renf_elem_overflows_d(renf_elem_t a, renf_t nf)
         fmpz_set_d(dbl_max, DBL_MAX);
         int cmp = renf_elem_cmp_fmpz(a, dbl_max, nf);
         fmpz_clear(dbl_max);
-        return cmp >= 1 ? 1 : 0;
+        return cmp > 0 ? 1 : 0;
     } else {
         renf_elem_neg(a, a, nf);
         int overflow = renf_elem_overflows_d(a, nf);

--- a/libeantic/src/renf_elem/get_d.c
+++ b/libeantic/src/renf_elem/get_d.c
@@ -30,7 +30,7 @@ static int renf_elem_overflows_d(renf_elem_t a, renf_t nf)
         fmpz_set_d(dbl_max, DBL_MAX);
         int cmp = renf_elem_cmp_fmpz(a, dbl_max, nf);
         fmpz_clear(dbl_max);
-        return cmp == 1 ? 1 : 0;
+        return cmp >= 1 ? 1 : 0;
     } else {
         renf_elem_neg(a, a, nf);
         int overflow = renf_elem_overflows_d(a, nf);

--- a/libeantic/srcxx/renf_elem_class.cpp
+++ b/libeantic/srcxx/renf_elem_class.cpp
@@ -15,6 +15,7 @@
 #include <cstdlib>
 #include <functional>
 #include <boost/numeric/conversion/cast.hpp>
+#include <boost/math/special_functions/sign.hpp>
 #include <boost/optional.hpp>
 #include <boost/lexical_cast.hpp>
 
@@ -146,7 +147,7 @@ bool relop_mpz(const renf_elem_class& lhs, const mpz_class& rhs, const int cmp)
     if (cmp == 0)
         ret = renf_elem_equal_fmpz(lhs.renf_elem_t(), r, lhs.parent().renf_t());
     else
-        ret = renf_elem_cmp_fmpz(lhs.renf_elem_t(), r, lhs.parent().renf_t()) == cmp;
+        ret = boost::math::sign(renf_elem_cmp_fmpz(lhs.renf_elem_t(), r, lhs.parent().renf_t())) == cmp;
     fmpz_clear_readonly(r);
 
     return ret;
@@ -175,7 +176,7 @@ bool relop_mpq(const renf_elem_class& lhs, const mpq_class& rhs, const int cmp)
     if (cmp == 0)
         ret = renf_elem_equal_fmpq(lhs.renf_elem_t(), r, lhs.parent().renf_t());
     else
-        ret = renf_elem_cmp_fmpq(lhs.renf_elem_t(), r, lhs.parent().renf_t()) == cmp;
+        ret = boost::math::sign(renf_elem_cmp_fmpq(lhs.renf_elem_t(), r, lhs.parent().renf_t())) == cmp;
     fmpq_clear_readonly(r);
 
     return ret;
@@ -239,8 +240,8 @@ bool relop_maybe_fmpz(const renf_elem_class& lhs, Integer rhs, const std::functi
     bool ret;
 
     maybe_fmpz(rhs,
-        [&](auto v) { ret = op(lhs.renf_elem_t(), v, lhs.parent().renf_t()) == cmp; },
-        [&](const fmpz_t v) { ret = renf_elem_cmp_fmpz(lhs.renf_elem_t(), v, lhs.parent().renf_t()) == cmp; });
+        [&](auto v) { ret = boost::math::sign(op(lhs.renf_elem_t(), v, lhs.parent().renf_t())) == cmp; },
+        [&](const fmpz_t v) { ret = boost::math::sign(renf_elem_cmp_fmpz(lhs.renf_elem_t(), v, lhs.parent().renf_t())) == cmp; });
 
     return ret;
 }
@@ -1048,11 +1049,11 @@ bool operator==(const renf_elem_class& lhs, long rhs) {
 }
 
 bool operator<(const renf_elem_class& lhs, long rhs) {
-    return renf_elem_cmp_si(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) == -1;
+    return renf_elem_cmp_si(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) < 0;
 }
 
 bool operator>(const renf_elem_class& lhs, long rhs) {
-    return renf_elem_cmp_si(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) == 1;
+    return renf_elem_cmp_si(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) > 0;
 }
 
 renf_elem_class& renf_elem_class::operator+=(unsigned long rhs)
@@ -1084,11 +1085,11 @@ bool operator==(const renf_elem_class& lhs, unsigned long rhs) {
 }
 
 bool operator<(const renf_elem_class& lhs, unsigned long rhs) {
-    return renf_elem_cmp_ui(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) == -1;
+    return renf_elem_cmp_ui(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) < 0;
 }
 
 bool operator>(const renf_elem_class& lhs, unsigned long rhs) {
-    return renf_elem_cmp_ui(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) == 1;
+    return renf_elem_cmp_ui(lhs.renf_elem_t(), rhs, lhs.nf->renf_t()) > 0;
 }
 
 renf_elem_class& renf_elem_class::operator+=(long long rhs)

--- a/libeantic/test/fmpz_poly_extra/t-isolate_real_roots.c
+++ b/libeantic/test/fmpz_poly_extra/t-isolate_real_roots.c
@@ -105,7 +105,7 @@ static void check_intervals(
 
         if ((j < n_exact) && (k < n_interval))
         {
-            if ((fmpq_cmp(exact + j, x) == 1) && (fmpq_cmp(y, exact + j) == 1))
+            if ((fmpq_cmp(exact + j, x) > 0) && (fmpq_cmp(y, exact + j) > 0))
             {
                 flint_printf("ERROR:\n");
                 flint_printf("the %wd-th exact root ", j);

--- a/libeantic/test/renf_elem/t-binop.cpp
+++ b/libeantic/test/renf_elem/t-binop.cpp
@@ -1,6 +1,6 @@
 /*
-    Copyright (C) 2017 Vincent Delecroix
-    Copyright (C) 2020 Julian Rüth
+    Copyright (C)      2017 Vincent Delecroix
+                  2020-2021 Julian Rüth
 
     This file is part of e-antic
 
@@ -102,7 +102,8 @@ TEST_CASE("Floor Division of renf_elem", "[renf_elem][binop][fdiv]")
     {
         fmpz_t& n = GENERATE_REF(take(4, fmpzs(state)));
 
-        if (!fmpz_is_zero(n)) {
+        if (!fmpz_is_zero(n) && !renf_elem_is_zero(a, nf))
+        {
             renf_elem_t b;
             renf_elem_init(b, nf);
             renf_elem_mul_fmpz(b, a, n, nf);
@@ -114,12 +115,45 @@ TEST_CASE("Floor Division of renf_elem", "[renf_elem][binop][fdiv]")
             renf_elem_fdiv(fdiv, b, a, nf);
 
             CAPTURE(fdiv, n);
-            if (renf_elem_is_zero(a, nf))
-                REQUIRE(fmpz_is_zero(fdiv));
-            else
-                REQUIRE(fmpz_equal(fdiv, n));
+            REQUIRE(fmpz_equal(fdiv, n));
 
             fmpz_clear(fdiv);
+            renf_elem_clear(b, nf);
+        }
+    }
+
+    SECTION("The quotient a // b when a / b is close to an integer")
+    {
+        fmpz_t& n = GENERATE_REF(take(4, fmpzs(state)));
+
+        renf_elem_t& error = GENERATE_REF(take(4, renf_elems(state, nf)));
+
+        renf_elem_div_ui(error, error, 1<<31, nf);
+
+        if (!fmpz_is_zero(n) && !renf_elem_is_zero(a, nf))
+        {
+            renf_elem_t b;
+            renf_elem_init(b, nf);
+            renf_elem_mul_fmpz(b, a, n, nf);
+            renf_elem_add(b, b, error, nf);
+
+            if (renf_elem_sgn(a, nf) == renf_elem_sgn(b, nf))
+            {
+                fmpz_t fdiv;
+                fmpz_init(fdiv);
+
+                CAPTURE(a, b, error);
+                renf_elem_fdiv(fdiv, b, a, nf);
+
+                CAPTURE(fdiv, n);
+                if (renf_elem_sgn(error, nf) == 0 || renf_elem_sgn(error, nf) == renf_elem_sgn(b, nf))
+                    REQUIRE(fmpz_cmp(fdiv, n) >= 0);
+                else
+                    REQUIRE(fmpz_cmp(fdiv, n) < 0);
+
+                fmpz_clear(fdiv);
+            }
+
             renf_elem_clear(b, nf);
         }
     }

--- a/libeantic/test/renf_elem/t-cmp.c
+++ b/libeantic/test/renf_elem/t-cmp.c
@@ -13,11 +13,18 @@
 
 #include "../../e-antic/renf_elem.h"
 
+static int sgn(int a) {
+  if (a == 0) return 0;
+  if (a < 0) return -1;
+  return 1;
+}
+
+
 static void check_cmp_fmpq(renf_elem_t a, fmpq_t b, renf_t nf, int ans)
 {
     int test;
 
-    if ((test = renf_elem_cmp_fmpq(a, b, nf)) != ans)
+    if (sgn(test = renf_elem_cmp_fmpq(a, b, nf)) != ans)
     {
         printf("FAIL:\n");
         printf("a = "); renf_elem_print_pretty(a, "x", nf, 10, EANTIC_STR_ALG & EANTIC_STR_D); printf("\n");
@@ -49,7 +56,7 @@ static void check_cmp(renf_elem_t a, renf_elem_t b, renf_t nf, int ans)
         abort();
     }
 
-    if ((test = renf_elem_cmp(a, b, nf)) != ans)
+    if (sgn(test = renf_elem_cmp(a, b, nf)) != ans)
     {
         printf("FAIL:\n");
         printf("a = "); renf_elem_print_pretty(a, "x", nf, 10, EANTIC_STR_ALG & EANTIC_STR_D); printf("\n");
@@ -58,7 +65,7 @@ static void check_cmp(renf_elem_t a, renf_elem_t b, renf_t nf, int ans)
         abort();
     }
 
-    if ((test = renf_elem_cmp(b, a, nf)) != -ans)
+    if (sgn(test = renf_elem_cmp(b, a, nf)) != -ans)
     {
         printf("FAIL:\n");
         printf("a = "); renf_elem_print_pretty(a, "s", nf, 10, EANTIC_STR_ALG & EANTIC_STR_D); printf("\n");

--- a/libeantic/test/renf_elem/t-cmp_fmpq.c
+++ b/libeantic/test/renf_elem/t-cmp_fmpq.c
@@ -11,6 +11,12 @@
 
 #include "../../e-antic/renf_elem.h"
 
+static int sgn(int a) {
+  if (a == 0) return 0;
+  if (a < 0) return -1;
+  return 1;
+}
+
 int main()
 {
     int iter;
@@ -40,7 +46,7 @@ int main()
             renf_elem_init(a, nf);
             renf_elem_set_fmpq(a, q1, nf);
 
-            if (renf_elem_cmp_fmpq(a, q1, nf) != 0 || renf_elem_cmp_fmpq(a, q2, nf) != fmpq_cmp(q1, q2))
+            if (renf_elem_cmp_fmpq(a, q1, nf) != 0 || sgn(renf_elem_cmp_fmpq(a, q2, nf)) != sgn(fmpq_cmp(q1, q2)))
             {
                 printf("FAIL:\n");
                 printf("a = "); renf_elem_print_pretty(a, "x", nf, 10, EANTIC_STR_ALG & EANTIC_STR_D); printf("\n");

--- a/libeantic/test/renf_elem/t-get_cfrac.c
+++ b/libeantic/test/renf_elem/t-get_cfrac.c
@@ -11,6 +11,12 @@
 
 #include "../../e-antic/renf_elem.h"
 
+static int sgn(int a) {
+  if (a == 0) return 0;
+  if (a < 0) return -1;
+  return 1;
+}
+
 static void check_nf_cf(renf_t nf, flint_rand_t state, const mp_bitcnt_t bits, const slong n, const slong num)
 {
     slong iter;
@@ -63,7 +69,7 @@ static void check_nf_cf(renf_t nf, flint_rand_t state, const mp_bitcnt_t bits, c
 
                 s = 0;
             }
-            if (ans != s)
+            if (sgn(ans) != s)
             {
                 printf("FAIL:\n");
                 printf("a = "); renf_elem_print_pretty(a, "x", nf, 10, EANTIC_STR_ALG & EANTIC_STR_D); printf("\n");

